flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/autoscaler/KubernetesScalingRealizerTest.java

```diff
@@ -30,12 +31,23 @@
 public class KubernetesScalingRealizerTest {
 
     @Test
-    public void testAutoscalerOverridesVertexIdsAreSorted() {
-
+    public void testAutoscalerOverridesStringDoesNotChangeUnlessOverridesChange() {
         KubernetesJobAutoScalerContext ctx =
                 TestingKubernetesAutoscalerUtils.createContext("test", null);
+        FlinkDeployment resource = (FlinkDeployment) ctx.getResource();
+
+        // Create resource with existing parallelism overrides
+        resource.getSpec()
+                .getFlinkConfiguration()
+                .put(PipelineOptions.PARALLELISM_OVERRIDES.key(), "a:1,b:2");
```

Comment:
Could we test `"a:1,b:2"` and `"b:2,a:1"`?

Also, the `assertThat` have  "a:1,b:2"`  as well, it's  better to extract  an field.

Reply:
There is no need to test any permutations since the overrides are deterministically set to `a:1,b:2` in the beginning. Then, new overrides are supplied via a LinkedHashMap which produces a deterministic `b:2,a:1`. Then the test verifies that the config string is not changed.

Reply:
Thanks for your clarification!

I prefer to test both of `"a:1,b:2"` and `"b:2,a:1"` for 2 reasons:

- We always use the black box testing, it means we don't care about some detailed logic or `Data Structure` inside of test method. We just test its feature.
- In the most of cases, the traversal order is determined when we using the `LinkedHashMap`, and the serialization result is often `b:2,a:1`, but I don't think the serialization result is fixed. That's means when 2 `LinkedHashMap` are equal, but the serialization results may not be equal. And that's why we need this PR, right?
  - So I don't think our test should make an assumption: the serialization result is fixed.
  - Also, the test is better to cover more cases. (For example, the `LinkedHashMap` is changed to `TreeMap` in the future.)


BTW, current change is fine, and this comment isn't necessary for me, so I approved it. If you think it's not needed, feel free to merge it. Thanks~

Reply:
Thanks for elaborating further on this. Testing is somewhat of an art. There is also an argument for striking a balance between a specific test and one which tests all possible code paths (maybe even ones which don't yet exist).

Black box testing is one way of testing. White box testing is also a valid approach. It depends on the situation and it is also a matter of taste and how confident one feels that the test covers the most important scenarios.

I completely understand your thought, and I've changed the test method to address this concern.

Reply:
Thank you for the feedback! ❤️